### PR TITLE
Minimal enterprise auth manager

### PR DESCRIPTION
**Pull request by henriknyman**
A minimal enterprise auth manager based on Apache Shiro,
with a custom file user realm mapping to our flat auth file
plus a new custom flat file mapping roles to multiple users,
and a static mapping between predefined roles and permissions.

### CODE REVIEW

File: community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java [hunk 1]
```diff
@@ -173,7 +169,7 @@ public void setUserPassword( String username, String password ) throws IOExcepti
 
         if ( existingUser.credentials().matchesPassword( password ) )
```
Comment (OliviaYtterbrink) [inline community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java]: Check password policy here instead of in setPassword?

File: community/security/src/main/java/org/neo4j/server/security/auth/User.java [hunk 2]
```diff
@@ -94,6 +104,7 @@ public boolean equals( Object o )
     public int hashCode()
     {
         int result = name != null ? name.hashCode() : 0;
+        result = 31 * result + ( group != null ? group.hashCode() : 0);
```
Comment (craigtaverner) [inline community/security/src/main/java/org/neo4j/server/security/auth/User.java]: Why would the group be part of the hashCode? That seems to imply that changing a users group somehow changes the identity of the user?

Comment (craigtaverner) [inline community/security/src/main/java/org/neo4j/server/security/auth/User.java]: Or is this only used to decide if the line in the auth file should be updated and the current location of a group field in that file means we need to do this?

File: community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java [hunk 3]
```diff
@@ -72,21 +72,28 @@
     private String serialize( User user )
     {
         return join( userSeparator, user.name(),
+                user.group(),
```
Comment (craigtaverner) [inline community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java]: If group was previously not part of the auth file, perhaps we should not add it, but add the groups file? That would remove a potentially unclean intermediate state.

File: community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java [hunk 4]
```diff
@@ -59,18 +59,18 @@ public void shouldFindAndAuthenticateUserSuccessfully() throws Throwable
     {
         // Given
         final InMemoryUserRepository users = new InMemoryUserRepository();
-        final User user = new User( "jake", Credential.forPassword( "abc123" ), false );
+            final User user = new User( "jake", "admin", Credential.forPassword( "abc123" ), false );
```
Comment (craigtaverner) [inline community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java]: I'm more comfortable having the tests also reflect the design of creating users and groups separately and assigning users to groups, instead of a group as a field of a user.

Comment (craigtaverner) [inline community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java]: Since this is the community auth manager, we could also argue that there is no concept of groups here.

File: community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java [hunk 5]
```diff
@@ -77,7 +83,7 @@ public void shouldStoreAndRetriveUsersByName() throws Exception
     {
         // Given
         FileUserRepository users = new FileUserRepository( authFile, NullLogProvider.getInstance() );
-        User user = new User( "jake", Credential.INACCESSIBLE, true );
+        User user = new User( "jake", "admin", Credential.INACCESSIBLE, true );
```
Comment (craigtaverner) [inline community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java]: As above, I feel like in community there is no concept of groups. Should we be testing for this here?

File: enterprise/NOTICE.txt [hunk 6]
```diff
@@ -1,29 +1,26 @@
...
 
 This product includes software ("Software") developed by Neo Technology.
 
-The copyright in the bundled Neo4j graph database (including the
-Software) is owned by Neo Technology. The Software developed and owned
-by Neo Technology is licensed under the GNU AFFERO GENERAL PUBLIC
-LICENSE Version 3 (http://www.fsf.org/licensing/licenses/agpl-3.0.html)
-("AGPL") to all third parties and that license, as required by the AGPL,
-is included in the LICENSE.txt file.
+The software ("Software") is developed and owned by Network Engine
```
Comment (craigtaverner) [inline enterprise/NOTICE.txt]: Should this be part of a security PR? Perhaps it should be at least in a separate commit?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedGroupsAccountBuilder.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    private Set<String> getRolesForGroup( String group )
+    {
+        switch ( group )
+        {
+        case "admin":
+        case "architect":
+        case BasicAuthManager.DEFAULT_GROUP:
+            return Collections.singleton( ShiroAuthSubject.SCHEMA_READ_WRITE );
```
Comment (craigtaverner) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedGroupsAccountBuilder.java]: Here we seems to be using the term 'Role' to mean Permission. I thought 'Role' was 'Group'. We should probably clarify and correct to match the current designs.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        assertAuthEnabled();
+        return realm.newUser( username, DEFAULT_GROUP, initialPassword, requirePasswordChange );
+    }
+
+    public User newUser( String username, String group, String initialPassword, boolean requirePasswordChange ) throws
+            IOException,
+            IllegalCredentialsException
+    {
+        assertAuthEnabled();
+        return realm.newUser( username, group, initialPassword, requirePasswordChange );
```
Comment (craigtaverner) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java]: I'm still uncomfortable with this hard linking of a user to a single group.

File: community/security/src/main/java/org/neo4j/server/security/auth/User.java (hunk 2 above)
Comment (henriknyman) [inline community/security/src/main/java/org/neo4j/server/security/auth/User.java]: Equivalence of User _objects_ is only used to detect concurrent modification of the auth repository  (throwing `ConcurrentModificationException`), so this `hashCode` just matches `equals`.
Whenever the user identity is compared it is by `user.name()`.

File: enterprise/NOTICE.txt (hunk 6 above)
Comment (henriknyman) [inline enterprise/NOTICE.txt]: Well, this was what the licensing plugin changed, and otherwise we did not pass the Team City tests. I guess there was a recent update to the licence? It is a bit weird why we would have to change this though... 

It is in a separate commit.

File: community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java (hunk 3 above)
Comment (henriknyman) [inline community/security/src/main/java/org/neo4j/server/security/auth/UserSerialization.java]: Last time we discussed this you agreed that it would make sense to keep the Unixy way of a default group in the user file for convenience, and then add a group file. I am not sure what you mean by unclean intermediate state?

File: community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java (hunk 4 above)
Comment (henriknyman) [inline community/security/src/test/java/org/neo4j/server/security/auth/BasicAuthManagerTest.java]: We are no actually testing the group, it is just passing something in.

We already have a plan to add multiple groups as a separate card. This is just the minimal first iteration.

File: community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java (hunk 5 above)
Comment (henriknyman) [inline community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java]: We are not testing the group. The reason group is in community is for backwards compatibility of the auth file and to simplify the architecture. The group is ignored and has no meaning to the community auth manager.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java (hunk 8 above)
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/ShiroAuthManager.java]: We have a card to change this :)

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedGroupsAccountBuilder.java (hunk 7 above)
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/PredefinedGroupsAccountBuilder.java]: Yes this pre-dates the latest design documents. I had a sketch of how to map the resource permission terminology we drew up in the last design document to Shiro roles and wildcard permission, but was waiting for it to stabilize.

File: community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java (hunk 1 above)
Comment (henriknyman) [inline community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthManager.java]: The password policy functionality is duplicated in REST UserService, with its own expected status code, hence the division of a setUserPassword (it is the old API renamed)
Here is an attempt to remove password policy from UserService and move validation into setUserPassword (but it changes the expected status code of one test..): https://github.com/henriknyman/neo4j/commit/4d79ced5def19fd9ea9ac599d050b371b2fa3f32

File: community/security/src/main/java/org/neo4j/server/security/auth/User.java (hunk 2 above)
Comment (craigtaverner) [inline community/security/src/main/java/org/neo4j/server/security/auth/User.java]: If we keep groups outside auth, then concurrent modification and identity would be the same.

File: community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java [hunk 9]
```diff
@@ -233,14 +239,18 @@ public void shouldIgnoreInvalidEntries() throws Throwable
         AssertableLogProvider logProvider = new AssertableLogProvider();
         Files.createDirectories( authFile.getParent() );
         Files.write( authFile, UTF8.encode(
-                "neo4j:fc4c600b43ffe4d5857b4439c35df88f:SHA-256,A42E541F276CF17036DB7818F8B09B1C229AAD52A17F69F4029617F3A554640F,FB7E8AE08A6A7C741F678AD22217808F:\n" ) );
+                "neo4j:admin:fc4c600b43ffe4d5857b4439c35df88f:SHA-256," +
```
Comment (craigtaverner) [inline community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java:242]: Is this a residual group field? Should be moved, I assume?

File: community/server/src/test/java/org/neo4j/server/rest/dbms/UserServiceTest.java [hunk 10]
```diff
@@ -200,25 +226,22 @@ public void shouldReturn404WhenChangingPasswordIfDifferentUser() throws Exceptio
     }
 
     @Test
-    public void shouldReturn404WhenChangingPasswordIfUnknownUser() throws Exception
+    public void shouldReturn422WhenChangingPasswordIfUnknownUser() throws Exception
```
Comment (craigtaverner) [inline community/server/src/test/java/org/neo4j/server/rest/dbms/UserServiceTest.java:229]: I thought the idea was to standardize on less informative return codes?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseAuthManagerFactory.java [hunk 11]
```diff
@@ -43,9 +50,30 @@ public EnterpriseAuthManagerFactory()
...
+        File authStoreDir = config.get( GraphDatabaseSettings.auth_store_dir );
+        File userStoreFile;
+        if ( authStoreDir != null )
+        {
+            userStoreFile = new File( authStoreDir, USER_STORE_FILENAME );
+        }
+        else
+        {
+            // Fallback on the directory of the legacy setting
+            userStoreFile = config.get( GraphDatabaseSettings.auth_store );
```
Comment (craigtaverner) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseAuthManagerFactory.java:63]: Should we support legacy settings? Perhaps since this is new code, we should not. Or do we have an intermediate state where current configs have no default?

Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseAuthManagerFactory.java:63]: Current config have no default, so we are using the existing default directory of the old setting. We need to revisit this after discussing with operability. Created a card for tracking it.

File: community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java (hunk 9 above)
Comment (henriknyman) [inline community/security/src/test/java/org/neo4j/server/security/auth/FileUserRepositoryTest.java:242]: This has slightly wrong name/assertions/error messages. Changed the test in a new PR to reflect the actual behaviour change (we now fail hard on a wrongly formatted file).

File: community/server/src/test/java/org/neo4j/server/rest/dbms/UserServiceTest.java (hunk 10 above)
Comment (henriknyman) [inline community/server/src/test/java/org/neo4j/server/rest/dbms/UserServiceTest.java:229]: 422 is what we are doing for the other types of illegal requests so you could argue it is less informative (no distinction in error code). We do however still provide exact descriptions in the messages.
I think standardising error message responses across REST and Bolt should be done as a separate PR.
